# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmDebugFrame.cpp
# Commit: d6506b5ee07b
# Full Hash: d6506b5ee07b5c65f889ce7c2214b0172f205898
# Author: Dmitry Bezhetskov <dbezhetskov@igalia.com>
# Date: 2021-09-16 09:43:03
# Regressor Bug: 1639153
# File Overlap Count: 1
# Description:
#   Bug 1639153 - Introduce indirect stubs to optimize call_indirect. r=lth
#   
#   This patch introduces indirect stubs.
#   An indirect stub is a stub that takes care of any switching activities needed for call_indirect.
#   Before this patch, we have to conservatively assume that any call_indirect's target can be from a foreign instance,
# ==============================================================================

diff -r 80a164c1054e -r d6506b5ee07b js/src/wasm/WasmDebugFrame.cpp
--- a/js/src/wasm/WasmDebugFrame.cpp	Wed Sep 15 11:02:57 2021 +0000
+++ b/js/src/wasm/WasmDebugFrame.cpp	Wed Sep 15 11:10:15 2021 +0000
@@ -34,8 +34,12 @@
 DebugFrame* DebugFrame::from(Frame* fp) {
   MOZ_ASSERT(
       GetNearestEffectiveTls(fp)->instance->code().metadata().debugEnabled);
-  auto* df =
-      reinterpret_cast<DebugFrame*>((uint8_t*)fp - DebugFrame::offsetOfFrame());
+  size_t offset = DebugFrame::offsetOfFrame();
+  if (fp->callerIsTrampolineFP()) {
+    offset +=
+        FrameWithTls::sizeWithoutFrame() + IndirectStubAdditionalAlignment;
+  }
+  auto* df = reinterpret_cast<DebugFrame*>((uint8_t*)fp - offset);
   MOZ_ASSERT(GetNearestEffectiveTls(fp)->instance == df->instance());
   return df;
 }